Tutorial: Let's talk to Kafka. How to send and receive application data from Apache Kafka®? (#1733)
aiven/devportal: docs/tutorials/talk-to-kafka.rst

        co1(kafka consumer co1)
        subgraph consumer group A
        co0
        co1
        end
        p0 -->co1

Apache Avro™ to produce and consume messages
---------------------------------------------

The Kafka brokers understand data as stream of bytes so one needs to pick a serializer and deserializer to convert the bytes into meaningful messages. 
Any format would do as long as it's consistent. For this tutorial, we're selecting Apache Avro which is an open-source project and one of the most popular serialization format.
Avro is defined by a schema and the schema is written in JSON. You can consider Avro as JSON with a schema attached to it.

Setting up a consumer to listen to Avro messages
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Here's an example of a Kafka consumer in Python using the `confluent-Kafka-python <https://github.com/confluentinc/confluent-kafka-python>`_ library to consume Avro-encoded messages.
On a terminal window, run the following consumer code. You won't see anything happening yet since there's no message to consume yet. 
Keep the program running and you'll create the producer program in the next section.

With SSL authentication:

> .. code:: python

Author: Made the change for producer/consumer swap. I'll be running the entire tutorial on a fresh VM now. then I can check if we need the learner to install any Python libraries.

> .. code:: python

Author: Just verified; you're right. Added `fastavro` requirement via https://github.com/aiven/devportal/pull/1733/commits/c039a69367d93a036e2aedf6b29a4e6f9a4c51d0.

        'client.id': '0001',
        'acks': 1,
        'compression.type': 'gzip'
    }

    # Configure the Avro schema registry
    schema_registry_conf = {'url': f"{SCHEMA_REGISTRY_URI}"}
    schema_registry_client = SchemaRegistryClient(schema_registry_conf)

    # Define the Avro schema for the message
    schema_str = """
    {
    "namespace": "example.avro",
    "type": "record",
    "name": "User",
    "fields": [
        {"name": "name", "type": "string"},
        {"name": "age",  "type": ["int", "null"]}
    ]
    }
    """

    # Create the Avro serializer
    avro_serializer = AvroSerializer(schema_str, schema_registry_client)

    # Create the Kafka producer
    producer = Producer(conf)

    # Define the message key and payload
    key = 'key1'
    payload = {'name': 'John', 'age': 30}

    # Serialize the message payload in Avro format
    serialized_payload = avro_serializer(key, payload)

    # Send the message to the Kafka topic
    producer.produce(topic='demo-topic', value=serialized_payload)

    # Wait for any outstanding messages to be delivered and delivery reports to be received
    producer.flush()

Observation
~~~~~~~~~~~

> In the above examples, we're using ``confluent-kafka-python`` library to send and consume Avro messages to/from a Kafka broker. You can follow the SSL protocol example to create producer and consumer codes for SASL protocol. Here's an overview of what each program does:

Reviewer: Why are you mentioning SASL here? should we give a hint on what they need to change for sASL?

Author: Good catch. Removed mention and added suggestion.

Tutorial: Let's talk to Kafka. How to send and receive application data from Apache Kafka®?
==============================================================================================

.. Note::

    This tutorial assumes basic knowledge on Apache Kafka.

Learning objectives
--------------------

- Basic overview of Apache Kafka and creating a highly available Apache Kafka service
- The concept of topic, partition, producer, consumer, and consumer groups 
- The need for data serialization in Kafka and use of Apache Avro™ to produce and consume messages

Overview
--------

> Getting started with Apache Kafka can be simple as you download the binary and start the bootstrap server. However, getting a production-ready Kafka cluster stand up with security and high-availability is a different story. 

Reviewer: I would avoid `simple`
```suggestion
Getting started with Apache Kafka can be straightforward: you download the binary and start the bootstrap server. However, getting a production-ready Kafka cluster stand up with security and high-availability is a different story. 
```

> If your goal is to use Kafka to benefit your application or data needs and NOT to learn Kafka administration, you can consider using `a managed Kafka service <https://aiven.io/kafka>`_.

Reviewer: ```suggestion
If your goal is to use Kafka to benefit your application or data needs and NOT to learn Kafka administration, you can consider using `a managed Kafka service <https://aiven.io/kafka>`_, allowing you to start producing and consuming data in minutes.
```

In this tutorial, we will learn how to create a highly available Apache Kafka service, go over the common tasks of producing and consuming messages, and finally use the popular Apache Avro specification to communicate with your Kafka service. 
This tutorial will use Python programming language.
Under the hood, the Python library will make use of the `Producer API <https://kafka.apache.org/documentation>`_ and the `Consumer API <https://kafka.apache.org/documentation>`_.

Prerequisites
-------------

To get started, you'll need:

- A Kafka service - either local or managed 
- `Python installed <https://www.python.org/downloads/>`_
- `Kafka-python <https://github.com/dpkp/kafka-python>`_ library installed, you can install it with the following command:

  .. code:: bash

    pip install kafka-python

> - `confluent-Kafka <https://github.com/confluentinc/confluent-kafka-python>`_ library installed, you can install it with the following command:  

Reviewer: Why are we using both kafka python and confluent kafka?

```suggestion
- `confluent-Kafka <https://github.com/confluentinc/confluent-kafka-python>`_ library installed that you'll use to produce messages in AVRO format, you can install it with the following command:  
```

  .. code:: bash

    pip install confluent-kafka

- `fastavro <https://github.com/fastavro/fastavro>`_ library installed, you can install it with the following command:

  .. code:: bash

    pip3 install fastavro
    
If you already have a Kafka service, jump to :ref:`tutorial-kaka-python-create-a-topic` section. 

Else, `sign up for an Aiven account <https://console.aiven.io/signup>`_ and follow the tutorial to create a Kafka service on Aiven.

Create an Aiven for Apache Kafka® service
-----------------------------------------

To create an :doc:`Aiven for Apache Kafka </docs/products/kafka>` service, you need to:

1. Log in to the `Aiven web console <https://console.aiven.io/>`_.
2. On the *Services* page, click **Create a new service**.

   This opens a new page with the available service options.

   .. image:: /images/platform/concepts/console_create_service.png
      :alt: Aiven Console view for creating a new service

3. Select **Apache Kafka®**.

4. Select the cloud provider and region that you want to run your service on.

> 5. Select `business-4` as service plan.

Reviewer: Why do we need a `business-4`? If there's a particular requirement (like being able to enable Karapace) we should mention, otherwise we should select a minor plan

5. Enter ``demo-kafka`` as name for your service.

6. Click **Create Service** under the summary on the right side of the console.

> The blinking blue icon besides your service name will indicate that VMs are being provisioned. Once the icon turns solid green, your Apache Kafka service is up and running.

Reviewer: ```suggestion
The blinking blue icon besides your service name will indicate that service is being provisioned. Once the icon turns solid green, your Apache Kafka service is up and running.
```

     #. Note the *Password* required for the SASL, we'll need it for authentication

* Create the keystore ``client.keystore.p12`` and truststore ``client.truststore.jks`` by following  :doc:`our article on configuring Java SSL to access Kafka <../products/kafka/howto/keystore-truststore>`

.. Warning::

  In the below examples, we just pass the name of the keystore and truststore files, but in actual use, the full path should be used.

You can also use the `Aiven command line tool <https://docs.aiven.io/docs/tools/cli.html>`_ to download the files. See the documentation for `avn service user-creds-download <https://docs.aiven.io/docs/tools/cli/service/user.html#avn-service-user-creds-download>`_

Details on the Aiven for Apache Kafka configuration can be found under the :ref:`kafka-tutorial-reference` section.

.. _tutorial-kaka-python-create-a-topic:

Create a topic
---------------

A topic in Kafka is a named stream of records that is stored within a Kafka cluster. Let's create a Kafka topic. 

If you have a local Kafka instance running, the command to create the topic might be something like this:

.. code:: bash

>     bin/kafka-topics.sh --create --topic demo-topic --bootstrap-server localhost:9092

Reviewer: We should mention that `localhost:9092` needs to be replaced with the server name

For an Aiven for Apache Kafka service, you can create the topic from the Aiven console. 

> From the **Topics** tab or the Kafka service overview page, click **Add topic**. Give the topic a name "demo-topic". Click **Add topic**.

Reviewer: ```suggestion
From the **Topics** tab or the Kafka service overview page, click **Add topic**. Give the topic a name ``demo-topic``. Click **Add topic**.
```

> Once this topic is created, we can see that the number of partitions is 1. 

Reviewer: ```suggestion
Once this topic is created, we can see that the default number of partitions is 1. 
```

Reviewer: where do we see the # of partitions?

> From the **Topics** tab or the Kafka service overview page, click **Add topic**. Give the topic a name ``demo-topic``. Click **Add topic**.

Reviewer: Can we add a screenshot here?

Reviewer: And also can we have this as list of steps in different dots?

> Once this topic is created, we can see that the default number of partitions is 1. 

Reviewer: Screenshot?

The concept of consumer group and consuming messages on Kafka
------------------------------------------------------------------

> Consumer group is the logical grouping of consumers. In Kafka, the consumer(s) must belong to a consumer group, even if it's the default consumer group. 

Reviewer: ```suggestion
Consumer group is the logical grouping of consumers. In Kafka, the consumer(s) must belong to a consumer group; if you don't define one, a random consumer group will be generated. 
```

For a Kafka cluster with multiple nodes, consumers within the same consumer group can exist on different nodes. 

.. mermaid::

    graph TD

        A(Topic) -- Partition 3 --> D[/Consumer 3/]
        A(Topic) -- Partition 4 --> E[/Consumer 4/]
        subgraph Consumer Group 2
        D
        E
        end

        A(Topic) -- Partition 1 --> B[/Consumer 1/]
        A(Topic) -- Partition 2 --> C[/Consumer 2/]
        subgraph Consumer Group 1
        B
        C

>         end

Reviewer: I believe this is wrong, since it seems that is splitting the topic across consumers belonging to two different consumer groups

Author: Thanks for the feedback. Fix is via https://github.com/aiven/devportal/pull/1733/commits/4c4cb085be4be48c0e95b7360ff8deec720d9a44.

                    " age: {}\n".format(key.id, user.name, user.age)
                )

        except KeyboardInterrupt:
                break

        # Commit the offset for the message
        consumer.commit()

Observation
~~~~~~~~~~~

Once you execute the consumer program, you should see the following output on the second terminal while the consumer program stays running until you hit ``ctrl+C``:

.. code:: shell

    Key --> 1
    User record --> name: John, age: 30

Here's an overview of what each program does:

Producer program
""""""""""""""""""

The producer program uses the ``confluent-kafka`` library to create an AvroProducer instance and send Avro messages to a Kafka broker. Here's a breakdown of what the program does:

- Import the necessary libraries: ``confluent_kafka``, ``confluent_kafka.avro``, ``ssl``, and ``io``.
- Define the URL of the Kafka broker and the Avro schema for the message.
- Set up SSL configuration by creating an SSL context and loading the client certificate, key, and CA certificate.
- Define the AvroProducer configuration, including the SSL settings and the Avro schema.
- Create an AvroProducer instance.

> - Send a sample Avro message to the Kafka broker.

Reviewer: I believe we need to show the parts of the code which are in charge of this. The code is long and we need to link back to what we're doing.

Also we need to add the fact that we define the schema

Author: Thanks @ftisiot 

I've addressed the other comments. A question regarding adding text regarding Avro piece by highlighting the code blocks. Will it be helpful if I add comments within the code blocks OR do you suggest I add the text before the actual code block?

Reviewer: That is a very good suggestion!

Author: https://github.com/aiven/devportal/pull/1733/commits/7abb8dab7dd9f9da81b4de816f68404c0a2e3b9d brings in comments within Avro code blocks to explain what each of the segments do. Please check if this looks Ok 🙏🏽

Let's talk to Kafka. How to send and receive application data from Apache Kafka®?
===================================================================================

.. Note::

    This tutorial assumes basic knowledge on Apache Kafka.

Learning objectives
--------------------

- Basic overview of Apache Kafka and creating a highly available Apache Kafka service
- The concept of topic, partition, producer, consumer, and consumer groups 
- The need for data serialization in Kafka and use of Apache Avro™ to produce and consume messages

Overview
--------

> Getting started with Apache Kafka can be straightforward: you download the binary and start the bootstrap server. However, getting a production-ready Kafka cluster stand up with security and high-availability is a different story. 

Author: This is an insightful comment. Please check if the change via https://github.com/aiven/devportal/pull/1733/commits/fd8736d0b8549ae88d7c142cf8ea11df4f094389 make sense.

> If your goal is to use Kafka to benefit your application or data needs and NOT to learn Kafka administration, you can consider using `a managed Kafka service <https://aiven.io/kafka>`_, allowing you to start producing and consuming data in minutes.

Author: Added change via https://github.com/aiven/devportal/pull/1733/commits/fd8736d0b8549ae88d7c142cf8ea11df4f094389.

          producer.send("demo-topic", message.encode('utf-8'))
          print(f"Message sent: {message}")
          time.sleep(1)

        producer.close()

Set up a consumer instance to start listening for messages
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
With the producer program executed, now run the consumer program for your chosen protocol. 

With SSL authentication:

.. code:: python

        # Import the required library
        from kafka import KafkaConsumer

        # Configuring the Kafka consumer  
        consumer = KafkaConsumer(
            "demo-topic",
            auto_offset_reset="earliest",
            bootstrap_servers=f"{HOST}:{SSL_PORT}", # From the connection information for managed service
            group_id="demo-consumer-group",
            security_protocol="SSL",
            ssl_cafile="ca.pem", # From the connection information for managed service
            ssl_certfile="service.cert", # From the connection information for managed service
            ssl_keyfile="service.key", # From the connection information for managed service
        )

        # Continuously poll for new messages
        for msg in consumer:

>           print("Message: ", msg.value)

Author: Good call. I added these lines under the **Observation** section. Let me know if you think there's a better place for this explanation.

>           print("Message: ", msg.value)

Author: Added change via https://github.com/aiven/devportal/pull/1733/commits/fd8736d0b8549ae88d7c142cf8ea11df4f094389.

        # Choose an appropriate SASL mechanism, for instance:
        SASL_MECHANISM = 'SCRAM-SHA-256'

        # Configuring the Kafka consumer 
        consumer = KafkaConsumer(
            "demo-topic",
            auto_offset_reset="earliest",
            bootstrap_servers = f'{HOST}:{SASL_PORT}', # From the connection information for managed service
            group_id="demo-consumer-group",
            sasl_mechanism = SASL_MECHANISM,
            sasl_plain_username = SASL_USERNAME, # From the connection information for managed service
            sasl_plain_password = SASL_PASSWORD, # From the connection information for managed service
            security_protocol = "SASL_SSL",
            ssl_cafile = "ca.pem" # From the connection information for managed service
        )

        # Continuously poll for new messages
        for msg in consumer:
          print("Message: ", msg.value.decode("utf-8"))

Observation
"""""""""""

You might have noticed ``key_deserializer``, ``key_serializer``, ``value_deserializer``, and ``value_serializer`` in these programs. Since Kafka brokers don't know about the records and only deal in bytes, the programs need to serialize 

> and deserialize data before making sense of them. 

Author: Good callout. This is a residue from a previous version of the code where these variables were used. I took them out but didn't notice this line. I'm taking this line out now.

To see some action on this terminal, run the producer code in another terminal. You will see the same record appear on the producer program terminal.

What would happen if there were two partitions in this case, ``p0`` and ``p1``? In this case, messages would be published to partition randomly. The consumer ``co0`` would take a round robin approach when consuming messages from this topic.

1 topic : 1 partition : 2 consumers : 1 consumer group
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Let's take a look at this setup where there are two consumers. ``co0`` and ``co1`` are registered to the same `consumer group A`. 

In this setup, one consumer will be sitting idle. This highlights an important concept in Kafka - records are processed in parallel and same partition cannot be assigned to multiple consumers in the same consumer group.

.. mermaid::
    
    graph LR;

        pr0(kafka producer pr0) -->p0(partition p0);
        subgraph topic
        p0
        end
        co0(kafka consumer co0)
        co1(kafka consumer co1)
        subgraph consumer group A
        co0
        co1
        end
        p0 -->co0 

> If the first consumer ``co0`` crashes for some reason, the other consumer ``co1`` in the consumer group will begin consuming messages from the last committed offset of the partition. 

Author: Did you mean "..cluster highly available."?

> If the first consumer ``co0`` crashes for some reason, the other consumer ``co1`` in the consumer group will begin consuming messages from the last committed offset of the partition. 

Author: Added change via https://github.com/aiven/devportal/pull/1733/commits/fd8736d0b8549ae88d7c142cf8ea11df4f094389

.. mermaid::
    
    graph LR;

        pr0(kafka producer pr0) -->p0(partition p0);
        subgraph topic
        p0
        end
        co0(CRASHED)
        co1(kafka consumer co1)
        subgraph consumer group A
        co0
        co1
        end
        p0 -->co1

Add schemas to messages with Apache Avro™
-------------------------------------------

The Kafka brokers understand data as stream of bytes so one needs to pick a serializer and deserializer to convert the bytes into meaningful messages. 
Any format would do as long as it's consistent. For this tutorial, we're selecting Apache Avro which is an open-source project and one of the most popular serialization format.
Avro is defined by a schema and the schema is written in JSON. You can consider Avro as JSON with a schema attached to it.

> Both of these programs use SSL authentication. You can follow the previous examples of this tutorial if you'd like to use SASL authentication instead.

Reviewer 2: ```suggestion
Both of these programs use SSL authentication. You can follow the previous examples in this tutorial if you'd like to use SASL authentication instead.
```

Setting up a producer to send Avro-encoded messages
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Here's an example of a Kafka consumer in Python using the `confluent-Kafka-python <https://github.com/confluentinc/confluent-kafka-python>`_ library to produce Avro-encoded messages.

> On a terminal window, run the following producer code. The program should execute and exit. You won't see anything yet since we need the consumer program to run next. 

Reviewer 2: ```suggestion
In a terminal window, run the following producer code. The program should execute and exit. You won't see anything yet since we need the consumer program to run next. 
```

        try:
            msg = consumer.poll(1.0)

            if msg is None:
                continue

            if msg.error():
                print("Consumer error: {}".format(msg.error()))
                continue
            
            user = avro_deserializer_value(
                msg.value(), SerializationContext(msg.topic(), MessageField.VALUE)
            )
            key = avro_deserializer_key(
                msg.key(), SerializationContext(msg.topic(), MessageField.KEY)
            )

            if user is not None and key is not None:
                print(
                    "Key --> {}\n"
                    "User record --> name: {},"
                    " age: {}\n".format(key.id, user.name, user.age)
                )

        except KeyboardInterrupt:
                break

        # Commit the offset for the message
        consumer.commit()

Observation
~~~~~~~~~~~

Once you execute the consumer program, you should see the following output on the second terminal while the consumer program stays running until you hit ``ctrl+C``:

.. code:: shell

    Key --> 1
    User record --> name: John, age: 30

Here's an overview of what each program does:

> Producer program

Reviewer 2: Put the content in this section (as in: copy and paste it) in L#380 where you actually define the producer

Author: With all the commits, the line# got all mixed up. This explanation is for the Avro part of the consumer/producer and is under the **Observation** section. The format is _I'm writing some code_ and then _here's what the code does_. 

Do you suggest I duplicate these texts to some places within the file? if yes, could you please comment the new line numbers? Thanks.

""""""""""""""""""

The producer program uses the ``confluent-kafka`` library to create an AvroProducer instance and send Avro messages to a Kafka broker. Here's a breakdown of what the program does:

- Import the necessary libraries: ``confluent_kafka``, ``confluent_kafka.avro``, ``ssl``, and ``io``.
- Define the URL of the Kafka broker and the Avro schema for the message.
- Set up SSL configuration by creating an SSL context and loading the client certificate, key, and CA certificate.
- Define the AvroProducer configuration, including the SSL settings and the Avro schema.
- Create an AvroProducer instance.
- Send a sample Avro message to the Kafka broker.

> Consumer program

Reviewer 2: Similar comment – put this content in the section you actually create the consumer in.

Author: Same response https://github.com/aiven/devportal/pull/1733#discussion_r1136459492
